### DISCUSSION TITLE
Pool/Node Governance

### DISCUSSION THREAD

**Discussion by arichard5** [cartesi/cips#5] [General]
Hello Cartesians,

There needs to be a frank discussion regarding how the pools and nodes are allowed to operate.

1. Decentralizing the network
2. Pool fees
3. Node running

Addressing the first issue can be straightforward.  Add a saturation limit to the pools and nodes, going over the saturation limit should adversely affect the block producing performance.  This will encourage delegators to look for pools that are not saturated and discourage massively funded private nodes.

Pool fees is a tougher topic but shouldn't take much effort to solve.  Instead of having two different fee structures, I propose that we incorporate the flat fee and the gas fee together.  Operators are free to choose their flat fee, this will let operators compete with each other, but will also have the gas fee paid through block production.  In this situation, operators are able to run their pool without losing money and then can add a percentage on top for profits as desired.  Pools that have a 0% flat fee will be kept in check by adding a pool saturation limit.  Once a pool hits a limit, they are free to create a new node for overflow delegators or delegators will choose another pool.

Ideally, all nodes would be pools.  This would foster a tighter community and allow for more competition among pool operators.  At the moment, this is not a feasible idea but it could be implemented over a few months time.

Looking forward to the communities thoughts.

dafodill-ii
yourctsimoon.eth

**Comment by TitusRyker** [comment]
Hello there,

Dafodill-ii, so we can talk on the same page here, I would like to ask what is your definition of decentralization and what is the point of decentralization. I understand decentralization to mean removing single/limited points of access to a network/community/activity/thing and removing singular/limited control/administration of any such network/community/activity/thing. The point of that is censorship resistance, stopping people from being denied access to open marketplaces and preventing the consolidation of power needed for one party to oppress another.

So please help me understand how adding a saturation cap to stake pools helps prevent censorship, helps people access the network/community, helps protect people, and/or removes a single point of access/control? I'll run through implementing a saturation limit to see and you tell me what I'm missing. I'll assume that whatever limit is chosen is low enough that some existing pools/nodes are over it as otherwise the limit would be meaninglessly high. That means the process starts by betraying those people who have been helping bootstrap the security of Noether by changing the rules on them. We should expect that some percentage of those who make up the pools/nodes over that limit leave the Cartesi community over this - thereby removing their value from the Cartesi ecosystem. It isn't just bug fish who make up the biggest delegators either; 4 of the 10 nodes with the most stake are pools. That means that you have people with small enough amounts of Cartesi that the gas fees are a big deal to them. Do you really want to tell someone with just a few hundred CTSI that they spent all that gas to get in a pool just to have the rules change on them and their ability to grow their CTSI decreased?

But probably not everyone who gets hit by the saturation limit will leave. They'll just spin up multiple nodes as seen on every other project I know of with a saturation limit (Cardano, Polkadot, etc). If we assume they keep the same amount of CTSI staked, they now represent an even higher portion of the network now that some have left. (And I suspect their stake would actually increase as they're given the opportunity to buy the dip created by those who leave selling off their CTSI.) Is this helping prevent censorship, helping people access the network/community or helping protect people? There is limiting the number of delegators who get slashed if only a subset of the nodes run by one of these multi-pool stakepool operators has a hiccup (an unlikely situation given the nodes are probably just replicated on the same hardware with the same internet connection & Etherium gateway meaning if one goes down they all go down). How else does the situation improve? 

What are the other downsides of pushing people to spin up multiple nodes? The one that comes to mind is staking operations are now more complicated and have more ETH transactions and therefore gas fees. (A greater number of nodes means more rebalance ops that need to be called.) In the end that's an increased cost that would have to be passed onto the delegators cutting into their yields and their motive for being with Cartesi. How else does the situation get worse?

Regarding the new fee structure, please clarify whether you are referring to starting the reward transaction by giving the node an amount of CTSI determined to be equal in value to that of the gas fee for the transaction or if you're referring to  having the transaction go out to some dex, exchange an amount of CTSI for ETH and then use that ETH to pay the gas fee. 

I'm interested to see how all nodes being pools/no nodes being solo-nodes can be implemented. I would feel rather safe in assuming that not all the existing solo-nodes will want to convert; so the only way to get rid of them would be to forcibly change the rules on them and make Cartesi no longer compatible with old nodes. I don't know if that's even possible, but if it is that's not an acceptable way to treat the people behind those solo-nodes. Even if this was done, say the nodes become stakepools; they disable delegation on the pools they create and the situation is effectively the same - you've just robbed them of some gas fees. Or, if they allow delegators on their node now those who want to provide a service to the Cartesi ecosystem by securing Noether and running a node have even more large pools to compete with. This makes it harder for new stakepools to come online and reach a sustainable level - directly hurting decentralization.

Just let those solo-node runners (who decided they wanted to be part of Cartesi before the dPOS system even had the d) continue doing their thing. They're not hurting anyone and have been key in the setup to provide security for Noether.

- Titus Ryker
(Transparency note: I run the Excelsior pool)

**Comment by tuler** [comment]
Personally I agree with @TitusRyker that capping doesn't look like a solution to decentralization.

Regarding the commission model you propose (percentage only above fees), I don't see how to implement it.

The ETH fee of the block production transaction is only known after the transaction is executed, not during the execution. We can't use the gas price of the transaction, as specified by the executor, and we only know the gas consumed up to that execution point, not the whole gas consumed. That's why in the gas based commission we use an oracle for the gas price (which can be different from the gas price of the transaction itself) and we let the owner define the gas amount. Not to mention EIP-1599, which makes it more complicated. The node also needs to pay for rebalance transactions and infrastructure costs (among other costs), so if the goal is to avoid unprofitable pools this is not enough.

**Comment by arichard5** [comment]
Hello Titus, Tuler

Thank you both for responding.

I had written a something of a reply before but had to work off-site today and was unable to finish my train of thought before it left the station. Hi-ho.

Capping may not be the solution to promote decentralization, can either of you think of a mechanism that would?  Re-staking your CTSI is a huge sink in terms of time and ETH.  I certainly wouldn't care to re-delegate to another pool for the current costs, however, is it appropriate that the network has 6 nodes that produce the majority of the blocks?

Now, I did come across a comment from Carlo where he stated "The reserve mining is only supposed to provide an initial incentive for node runners to join and bootstrap the Noether network" and "The services in the network themselves should provide the incentive to run nodes."  If this is the case, then all of this is moot and we should carry on status quo, not worrying about a balanced network until 2024, right?

I'll leave it here for the conversation's sake and look forward to one of your responses.

dafodill-ii

**Comment by augustoteixeira** [comment]
One possible solution for decentralizing pools would be to make the owner of the pool a smart contract that enforces several restrictions. We could implement a contract with a few "common sense" restrictions: like speed bumps, limits on the amount increased...

Then it would be up for the pool creators to chose whether they will restrict themselves in order to attract costumers or not. And it would be up for the stakers to decide if the restrictions are sensible for them.

What do you think?

**Comment by alexandre-abrioux** [comment]
Hi! I was talking with a colleague that runs a Fuse node the other day and he explained to me that they faced the same issue at the beginning, meaning pools were running with very low or even 0% fee, leading to stake centralization in the network, and they since instated a configurable global minimum fee in their protocol.

See: https://docs.fuse.io/general/fuse-network-blockchain/validators-and-delegation
> A delegator receives the share of Fuse block rewards proportionate to their stake minus the fee paid to their chosen validator. The minimum fee is currently set at 15%. This value was chosen to mitigate the risk of validator consolidation through offering very low delegation fees.

I found this solution very interesting. I was also thinking that the minimum fee could potentially be dynamically adjusted with current CTSI/ETH market price, instead of being manually modified from time to time by the Cartesi team— but this would make the implementation a little bit more difficult.

Just wanted to share this thought!

**Comment by tuler** [comment]
The “dynamically adjusted with current CTSI/ETH market price” is already implemented, by the gas tax commission model, which nobody is using.

**Reply by alexandre-abrioux** [reply]
Yes unfortunately... So yes for the gas commission model it would be easy to implement, just define a global minimum gas fee. For the percent commission model, we could reuse some of the code. We could determine a starting minimum fee related to current CTSI/ETH market price, let's say 10% for the sake of example. If pools configure a lower fee than 10%, the fee would remain 10% at block production. But when the CTSI/ETH market price increases, the global minimum fee would decrease accordingly with a linear slope or curved slope (TBD).

**Comment by arichard5** [comment]
Hi all,

Since the fees are variable by the operator, that isn't the issue.

In my opinion, the real issue is the consolidation of CTSI in relatively few pools and nodes.

I have seen a few delegators asking what is wrong with their pool as they have had no rewards in 1-2 months.  Yes, I know that given enough time all pools should generate the same rewards, however, delegators will not have the patience to wait their pool's turn to mint a block.  If we can resolve delegators choosing from 3 dominant pools, it will help the network decentralize greatly.

@TitusRyker, defending the solo-nodes is defending Binance and other large exchanges.  I think they can afford to not dominate every POS network.  I'm sure there are small independent solo-nodes but what benefit does a node bring if it isn't producing?

**Comment by alexandre-abrioux** [comment]
Hi @arichard5! Thanks for your reply. I think we are referring to two different issues:

1. Large nodes dominate the network / small node do have a very low probability of earning rewards.
2. Stakers are concentrating their delegation into 1-2 pools, giving too much power to those pools.

My proposition—to impose a minimum fee on pools—was to solve issue number (2).

Regarding issue number (1), I don't have a solution for this, and I doubt it is possible to solve. Individuals with high capital will always find ways to generate profit faster. In PoW or PoC they will invest in hardware, and in PoS they will stake aforementioned capital.

If we decide to decrease the probability of earning rewards when the stake on a node is too high, then those individuals will create more nodes and split their stake on those nodes. The issue is that those new nodes will physically run on the same server or cloud provider, which is not solving the centralization issue.

The only solution I see to counterbalance this is to get more and more people delegating CTSI, and offer them a fair delegation system so that the global delegated stake amount could be split onto as much nodes as possible, in order to decentralize the network as much as possible.
